src/core/core.scale.js

```diff
@@ -93,10 +83,117 @@ function getPixelForGridLine(scale, index, offsetGridLines) {
 	return lineValue;
 }
 
-function computeTextSize(context, tick, font) {
-	return helpers.isArray(tick) ?
-		helpers.longestText(context, font, tick) :
-		context.measureText(tick).width;
+/**
+ * Returns {width, height, offset} objects for the first, last, widest, highest tick
+ * labels where offset indicates the anchor point offset from the top in pixels.
+ */
+function computeLabelSizes(ctx, tickFonts, ticks, cache) {
```

Comment:
I would still refactor this (a lot)
 - get [toTextLines](https://github.com/chartjs/chartjs-plugin-datalabels/blob/c759a93a08924a0c8c4ef907fc3c3f0561960541/src/utils.js#L27) from datalabels to helpers.
 - extract garbage cleanup to separate function:
```js
function garbageCollect(caches, length) {
	helpers.each(caches, function(cache) {
		var gc = cache.gc;
		var gcLen = gc.length / 2;
		var i;
		if (gcLen > length) {
			for (i = 0; i < gcLen; ++i) {
				delete cache.data[gc[i]];
			}
			gc.splice(0, gcLen);
		}
	});
}
```
- refactor whats remainign:
```js
function computeLabelSizes(ctx, tickFonts, ticks, caches) {
	var length = ticks.length;
	var widths = [];
	var heights = [];
	var offsets = [];
	var i, j, jlen, label, tickFont, fontString, cache, lineHeight, width, height, widest, highest;

	for (i = 0; i < length; ++i) {
		label = helpers.toTextLines(ticks[i].label);
		tickFont = ticks[i].major ? tickFonts.major : tickFonts.minor;
		ctx.font = fontString = tickFont.string;
		cache = caches[fontString] || {data: {}, gc: []};
		lineHeight = tickFont.lineHeight;
		width = 0;
		for (j = 0, jlen = label.length; j < jlen; ++j) {
			width = helpers.measureText(ctx, cache.data, cache.gc, width, label[j]);
		}
		height = jlen * lineHeight;

		widths.push(width);
		heights.push(height);
		offsets.push(lineHeight / 2);
	}
	garbageCollect(caches, length);

	widest = widths.indexOf(Math.max.apply(null, widths));
	highest = heights.indexOf(Math.max.apply(null, heights));

	function valueAt(idx) {
		return {
			width: widths[idx] || 0,
			height: heights[idx] || 0,
			offset: offsets[idx] || 0
		};
	}

	return {
		first: valueAt(0),
		last: valueAt(length - 1),
		widest: valueAt(widest),
		highest: valueAt(highest)
	};
}
```

I did not test this, so there might be typo or two.

Reply:
I agree with most of this comment, but I would keep the `toTextLines` change for another PR because it introduces a new feature and it will be easier to review and keep track of it in a separate commit. I was thinking to do that right after this PR is merged.

Reply:
I like this refactoring. Made changes leaving the `toTextLines` related code as is.